### PR TITLE
ci: Build pre-release versions

### PR DESCRIPTION
**Pull request by BenjaminHalko**
Uses [semantic release](https://github.com/semantic-release/semantic-release/tree/master) to build prereleases from dev. (it is now used when making non-pre-release builds as well)

TODO:
- [x] Testing

### CODE REVIEW

**Comment by BenjaminHalko** [comment]
[Testing](https://github.com/BenjaminHalko/revanced-manager/actions/runs/7360174911) on my own fork seems to be successful.

**Comment by PalmDevs** [comment]
Does the built APK match the bumped version? Looking into this, it may be one version behind.

**Comment by BenjaminHalko** [comment]
I think you may be correct... `flutter build apk` should be a part of `.releaserc` and not the workflow, so that it gets executed after the version bump.

**Comment by Ushie** [comment]
@PalmDevs What were the complications we had when we wanted semantic releases?

**Comment by PalmDevs** [comment]
> I think you may be correct... `flutter build apk` should be a part of `.releaserc` and not the workflow, so that it gets executed after the version bump.

We can run semantic-release twice with different configs. The first time is run before building which is to bump pubspec.yaml. The second time is to finally upload.

**Comment by PalmDevs** [comment]
> @PalmDevs What were the complications we had when we wanted semantic releases? 

Same issue we currently have at the moment.

**Comment by Ushie** [comment]
How does this plan to handle version code and version name bumps?

**Comment by PalmDevs** [comment]
> How does this plan to handle version code and version name bumps?

It handles version names as is, version code will be incremented by one from what we currently have.

> Version names are decided by semantic release, but we went with a stupid design for version code and I don't think it incorporates dev version

The plugin in this PR bumps by one from what we currently have. Development versions will always be ahead or equal in version code unless someone decides to push to the main branch which we have restrictions for.

> We also need to answer the question of when does the version code bumps happen? 

Before the build process:

1. Bump everything and commit
2. Build APK
3. Upload, generate changelogs, and release

**Comment by Ushie** [comment]
> The plugin in this PR bumps by one from what we currently have. Development versions will always be ahead or equal in version code unless someone decides to push to the main branch which we have restrictions for. 

Every single build is gonna bump by 1, so dev 1 and dev 2 both bump the version codes I assume?

**Comment by PalmDevs** [comment]
> > The plugin in this PR bumps by one from what we currently have. Development versions will always be ahead or equal in version code unless someone decides to push to the main branch which we have restrictions for. 
> 
> Every single build is gonna bump by 1, so dev 1 and dev 2 both bump the version codes I assume? 

Yep, every trigger/build = 1 bump

**Comment by oSumAtrIX** [comment]
I remind here that ReVanced Manager is not a valid candidate for semantic release

**Comment by BenjaminHalko** [comment]
Ok, I think this may work now.

**Comment by BenjaminHalko** [comment]
Ok I think it works now!

File: .releaserc.bump.json
Comment (oSumAtrIX) [inline .releaserc.bump.json:1]: Whats the reason for separating the releaserc file to multiple?

Comment (BenjaminHalko) [reply .releaserc.bump.json:1]: It had to be split into 2 steps. 1 for getting the version. Then 1 for uploading after building.

**Review by oSumAtrIX — APPROVED** [review]

**Pull request merged**